src/queryIterator.ts

```diff
@@ -174,13 +177,41 @@ export class QueryIterator<T> {
     );
   }
 
-  private createQueryExecutionContext() {
-    return new ProxyQueryExecutionContext(
+  private createPipelinedExecutionContext(partitionedExecutionInfo: PartitionedQueryExecutionInfo) {
+    return new PipelinedQueryExecutionContext(
       this.clientContext,
+      this.resourceLink,
       this.query,
       this.options,
-      this.fetchFunctions,
-      this.resourceLink
+      partitionedExecutionInfo
     );
   }
+
+  private isPartitionedExecutionError(error: any): error is ErrorResponse {
+    return (
+      error.code === StatusCodes.BadRequest &&
+      "substatus" in error &&
+      error["substatus"] === SubStatusCodes.CrossPartitionQueryNotServable
+    );
+  }
+
+  private async executeMethod(methodName: Exclude<keyof IExecutionContext, "hasMoreResults">) {
+    try {
+      return await this.queryExecutionContext[methodName]();
+    } catch (err) {
+      if (this.isPartitionedExecutionError(err)) {
+        // if this's a partitioned execution info switches the execution context
+        const partitionedExecutionInfo = err.body.additionalErrorInfo;
+        this.queryExecutionContext = this.createPipelinedExecutionContext(partitionedExecutionInfo);
+        try {
+          // retry call with same method
+          return this.queryExecutionContext[methodName]();
+        } catch (e) {
+          throw e;
```

Comment:
do we need a try catch if we are just going to throw the error?

Reply:
You are correct. Will remove